src/app/util/af-main-common.cpp

```diff
@@ -335,129 +333,18 @@ static EmberStatus send(EmberOutgoingMessageType type, uint64_t indexOrDestinati
     return status;
 }
 
-EmberStatus emberAfSendMulticastWithAliasWithCallback(GroupId multicastId, EmberApsFrame * apsFrame, uint16_t messageLength,
-                                                      uint8_t * message, EmberNodeId alias, uint8_t sequence,
-                                                      EmberAfMessageSentFunction callback)
+EmberStatus emberAfSendUnicastWithCallback(DataModelContext & context, EmberApsFrame * apsFrame, uint16_t messageLength, uint8_t * message, EmberAfMessageSentFunction callback)
 {
-    apsFrame->groupId = multicastId;
-    return send(EMBER_OUTGOING_MULTICAST_WITH_ALIAS, multicastId, apsFrame, messageLength, message,
-                true, // broadcast
-                alias, sequence, callback);
-}
-
-EmberStatus emberAfSendMulticastWithCallback(GroupId multicastId, EmberApsFrame * apsFrame, uint16_t messageLength,
-                                             uint8_t * message, EmberAfMessageSentFunction callback)
-{
-    apsFrame->groupId = multicastId;
-    return send(EMBER_OUTGOING_MULTICAST, multicastId, apsFrame, messageLength, message,
-                true, // broadcast?
-                0,    // alias
-                0,    // sequence
-                callback);
-}
-
-EmberStatus emberAfSendMulticast(GroupId multicastId, EmberApsFrame * apsFrame, uint16_t messageLength, uint8_t * message)
-{
-    return emberAfSendMulticastWithCallback(multicastId, apsFrame, messageLength, message, NULL);
-}
-
-EmberStatus emberAfSendMulticastToBindings(EmberApsFrame * apsFrame, uint16_t messageLength, uint8_t * message)
```

Comment:
And _definitely_ do not remove this API!  AS you might have noticed in `reporting.cpp`, it's needed for reporting as currently implemented; changing from "send to all the multicast bindings" to "send to all the unicast bindings" is quite a big semantic change: it sends to a different sent of destinations!

Reply:
The thing which confuse me it that, eventually `SendMessage` of all types of message will be handled by `ExchangeContext::SendMessage`. There will be unicast exchanges, which send unicast messages, there will be multicast exchanges, which send multicast messages, and it is totally transparent to DataModel. The type is exchanges can be fetched using `ExchangeContext::IsUnicast` or `ExchangeContext::IsMulticast` or `ExchangeContext::IsBroadcast`. The type of exchange is decided when creating the exchange.

So do we really need separate APIs for each type of message ? If we keep these APIs, they will both become `ExchangeContext::SendMessage`

Reply:
I'm fine to add them back. They are not causing any trouble if we just ignore them.

Reply:
I think we should probably sort out the questions in https://github.com/project-chip/connectedhomeip/pull/3602#discussion_r526985171 before we get too far into the discussion here, but the way I see it, the operations the data model needs from the message layer are:

1. Respond to this incoming message, using the exchange it came in on.
2. Create a new exchange for a given node id, do some message sending/receiving on it.
3. Create a new exchange for a given group id, do some message sending (and maybe receiving) on it.
4. Tear down an exchange, whether we created it or whether it was created by an incoming message, once we expect to not send/receive any more messages on it.

> and it is totally transparent to DataModel

For items 2 and 3, data model needs to _create_ the exchange when it wants to start a new transaction, right?  That is what the APIs involved should do: data model code should be able to say "hey, I want to start a new multicast transaction" and we should have some single function or small set of functions (depending on how the target of the multicast is being identified: bindings vs raw group id), which does the work involved in that: setting up the new exchange and then sending the message.

Reply:
> I'm fine to add them back. They are not causing any trouble if we just ignore them.

OK, good.  Let's leave them for now, then.  :)